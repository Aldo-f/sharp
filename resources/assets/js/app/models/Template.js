import util from '../../util';

class Template {
    constructor(fieldKey, templateName) {
        this.fieldKey = fieldKey;
        this.name = templateName;
    }

    get name() { return this._name }
    get fieldKey() { return this._fieldKey }

    set name(name) {
        this._name=util.capitalize(name);
    }
    set fieldKey(fieldKey) {
        this._fieldKey=util.capitalize(fieldKey);
    }

    get compName() {
        return util.hyphenate(`SharpTemplate${this.name}${this.fieldKey}`);
    }

    get exists() {
        return !!Vue.options.components[this.compName];
    }

    get compNameOrDefault() {
        return this.exists ? this.compName : null;
    }

// static utils

    static isTemplateProp(fieldPropName) {
        return /^.+Template$/.test(fieldPropName);
    }

    static parseTemplateName(fieldPropName) {
        let result=/^(.+)Template$/.exec(fieldPropName);
        if(result.length>1)
            return result[1];
        return null;
    }
}

export default Template;