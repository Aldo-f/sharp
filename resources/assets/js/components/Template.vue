<template>
    <rendered-template :template="template"
                       :template-data="templateData"
                       :template-props="templateProps">
    </rendered-template>
</template>

<script>
    export default {
        name: 'SharpTemplate',

        components: {
            RenderedTemplate: {
                functional: true,
                render(h, { props }) {
                    const { template, templateProps, templateData } = props;
                    return h({
                            name: 'AAA',
                            template:`<div>${template}</div>`,
                            props: templateProps
                        }, {
                        props: templateData
                    });
                }
            }
        },

        props: {
            name: String,
            templateData: Object,
            template: String,
        },

        computed: {
            templateProps() {
                return Object.keys(this.templateData);
            },
        }
    }
</script>