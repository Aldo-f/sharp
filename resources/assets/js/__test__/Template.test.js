import Vue from 'vue/dist/vue.common.js';
import Template from '../components/Template.vue';



describe('template-component',()=>{
    Vue.component('sharp-template', Template);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-template name="Test" :template-data="{ name:'Antoine' }" template="<em v-text='name'></em>"></sharp-template>
            </div>
        `
    });

    it('can mount given template wrapped in div tag', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});


async function createVm(customOptions={}) {
    const vm = new Vue({
        el: '#app',
        mixins: [customOptions],
    });

    await Vue.nextTick();

    return vm.$children[0];
}