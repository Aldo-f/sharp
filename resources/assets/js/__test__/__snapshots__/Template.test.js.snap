// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template-component can mount Password field 1`] = `
"
            <div id=\\"app\\"><div><em>Antoine</em></div></div>
        "
`;
